### ISSUE TITLE
Accessibility: expose element reference

### ISSUE THREAD

**Issue by AutoSponge** [puppeteer/puppeteer#3641]
### Steps to reproduce

While using the puppeteer api `page.accessibility.snapshot()` the AXNode model leaves no trace of the DOM implementation. The underlying calls to [CDP](https://chromedevtools.github.io/devtools-protocol/tot/Accessibility#method-getFullAXTree) return `nodeId` (and `backendDOMNodeId`).

I suggest one or more of the following (will update if more ideas come in comments):
1. Give AXNode a new method 'asElement' which will create an `ElementHandle` from the stored nodeId or remoteObject reference.
1. Expose the nodeId or remoteObjectId on the AXNode (so it can be passed to other CDP calls).
1. Expose AXNode so CDP calls to `Accessibility.getPartialAXTree` can use this code to normalize the result the way `snapshot()` does.

* Puppeteer version: 1.10+
* Platform / OS version: any
* URLs (if applicable):
* Node.js version: latest

_Please include code that reproduces the issue._

```js
// ... open google.com
const tree = await this.page.accessibility.snapshot();
console.log(tree);
```

Yields: 

```js
{ role: 'WebArea',
  name: 'Google',
  children:
   [ { role: 'button', name: 'Dismiss' },
     { role: 'link', name: 'About' },
     { role: 'link', name: 'Store' },
     { role: 'link', name: 'Gmail' },
     { role: 'link', name: 'Images' },
     { role: 'button', name: 'Google apps' },
     { role: 'link', name: 'Sign in' },
     { role: 'combobox',
       name: 'Search',
       description: 'Search',
       focused: true,
       autocomplete: 'both' },
     { role: 'button', name: 'Search by voice' },
     { role: 'button',
       name: 'Google Search',
       description: 'Google Search' },
     { role: 'button',
       name: 'I\'m Feeling Lucky',
       description: 'I\'m Feeling Lucky' },
     { role: 'img', name: 'Google' },
     { role: 'link', name: 'Privacy' },
     { role: 'link', name: 'Terms' },
     { role: 'combobox', name: 'Settings', haspopup: 'menu' },
     { role: 'link', name: 'Advertising' },
     { role: 'link', name: 'Business' } ] }
```

But more useful would be:

```js
{ role: 'WebArea',
  name: 'Google',
  nodeId: 1,
  children:
   [ { role: 'button', name: 'Dismiss', nodeId: 4 },
     { role: 'link', name: 'About', nodeId: 6 },
     { role: 'link', name: 'Store', nodeId: 7 },
     { role: 'link', name: 'Gmail', nodeId: 8 },
     { role: 'link', name: 'Images', nodeId: 11 },
     { role: 'button', name: 'Google apps', nodeId: 47 },
     { role: 'link', name: 'Sign in', nodeId: 48 },
 // ...
     { role: 'link', name: 'Business', nodeId: 129 } ] }
```

If approved, I feel I could make the needed changes to submit a PR.

**Comment by FremyCompany** [comment]
I would love to see this implemented

**Comment by FremyCompany** [comment]
In case someone wanted to volunteer working on this, is there documentation and a few pointers that would help get started on this issue?

**Comment by wycats** [comment]
I am also interested in this.

**Comment by AutoSponge** [comment]
> I suggest one or more of the following...

I think I can make this happen. Just need to understand the preference of the choices I outlined above (or have a better choice). Community or maintainer preference.

**Comment by FremyCompany** [comment]
Both options `1` and `3` sound great to me

**Comment by JoelEinbinder** [comment]
Sorry I should have been on top of this discussion sooner. Thanks for the interest everyone!

I definitely don't want to directly expose CDP's nodeIds. Instead the API should deal in ElementHandles.  And because of the way we make element handles, the api needs to expose a function that returns a promise to an ElementHandle. Otherwise we'd need to figure out how memory management would work, which would certainly be a breaking change.

Right now the AXNode is returned as a serilizable JSON object. This is done to make clear this this object is inert. It contains the state of the AXNode at the time of the snapshot. I feel a little bit weird adding a function to this object. But I don't see a better way.

What should that function be called? `.node()` is a bit strange, because the thing you are calling it on is also a "node". `.DOMNode()` is technically correct. `.element()` feels the most natural to me, and is consistent with `ElementHandle` in that we mislabel DOMNodes as elements 😄.

Not all Accessibility Nodes are directly associated with a DOM node, for example the popup menu of a `<select>` element. Should this return `null`, or the ElementHandle for its parent select? And sometimes AXNode's are associated with a DOMNode within the user agent shadow DOM. For example, the text inside a textarea in Chromium is implemented as a text node inside a div in the shadow dom of the textarea. Should these shadowy elements be exposed? Or `null`? Or their shadow host?

**Comment by FremyCompany** [comment]
@JoelEinbinder Thanks for the good questions!

Trying to answer based on my own use case, I would love for AXNodes pointing to Shadow DOM subtrees to give me a reference to the element inside the subtree, even if it's a closed/native one. That would be my preference.

If that is too difficult to implement, and I would instead get a reference to the last author-inserted element, that would be okay by me in most cases though, and actually more convenient in a few. A couple of cases would be significantly more annoying, but I'm sure I could hardcode them, because there aren't many elements which are implemented with closed shadow trees in Blink as far as I know.

Regarding the name of the function, I'm not too concerned about this, and I think consistency with what you have today is a good way of narrowing this down, except if someone feels strongly about this.

**Comment by AutoSponge** [comment]
`asElement()` was what I saw originally in other, similar areas of puppeteer but `element()` is fine. Personally, the native shadow implementation isn't helpful as I'm trying to map issues with accessibility back to their author implementation--WebComponents included (e.g., how another developer will see it in the devtools, not how the engine sees it). If that's something others want, please provide a second method for the "shadow host || element".

So, I'll make the same argument about the example of `<select>`: to a developer in devtools, the scrim of an opened `<select>` is abstracted away (and also invisible to the accessibility tab). That information is less helpful than the `role` and `name` exposed to AT.

**Comment by JoelEinbinder** [comment]
`asElement` is actually our way to do casting. It does `x instanceOf ElementHandle ? this : null`.

@AutoSponge  I'm not sure what you mean about the scrim being hidden from the accessibility tab.
![Screenshot from 2019-04-19 09-43-44](https://user-images.githubusercontent.com/4624233/56433980-aa5f7380-6287-11e9-8f1e-a49c27c8f994.png)
I'm talking about the MenuListPopup here, which doesn't correspond to any element in DOM, shadow or otherwise.

@FremyCompany The accessibility engine doesn't know the difference beteween user agent shadow nodes and just regular DOM nodes. So it would require more effort to filter them out. However I can't think of a case where they would be useful to web developers. Why would you want to see them?
![image](https://user-images.githubusercontent.com/4624233/56434184-6751d000-6288-11e9-97c7-055202e056a3.png)

Almost every DOM element has user agent shadow dom, besides really basic ones like div and span. It is just an implementation detail of Chromium though.

**Comment by aslushnikov** [comment]
> I feel a little bit weird adding a function to this object. But I don't see a better way.

@JoelEinbinder I prefer an option that injects an `axNodeId` to every AXNode and a separate method to resolve element handle from this ID:

```js
const snapshot = await page.accessibility.snapshot({
  axNodeIds: true
});
const elementHandle = await page.accessibility.resolveElement(snapshot.axNodeId);
```

Alternatively, we can invest some effort and teach `ElementHandle` to work with `backendNodeId` internally instead of a remote object. This way the following will work without any memory concerns:

```js
const snapshot = await page.accessibility.snapshot({
  elementHandles: true
});
const elementHandle = snapshot.elementHandle;
```

**Comment by FremyCompany** [comment]
@JoelEinbinder Well, in my case I am trying to emulate a screenreader experience of some arbitrary web content, so being able to see internal controls of, say, a `<video controls>` element is rather useful. If you are not interested in those elements, it's not very difficult to walk up the tree until you find something in the light tree again, so I wouldn't worry about filtering them out.

**Comment by JoelEinbinder** [comment]
@FremyCompany If you are trying to emulate a screen reader experience, why do you need the HTML element at all? Everything should be contained within the accessibility tree.

**Comment by FremyCompany** [comment]
@JoelEinbinder Things are more complex than that :) For instance on Windows, there are APIs like `UIA TextPattern` which provides line-by-line text navigation and allows you to find out about the precise location of the screens of words (more generally any text range) which you cannot get from the accessibility tree but can get with things like getBoundingClientRect on Range etc... Also, many controls need to be interacted with, which means that you need to be able to ".click()" them etc which you cannot do from the AXNodes as far as I know.

**Comment by JoelEinbinder** [comment]
All of those things are exposed through the real ax tree from chrome. It’s not exposed in puppeteer just because it takes some effort and I didn’t get around to it. I can probably add them if you have use for them. I definitely want to add clicking.

However if you are building a screen reader on top of puppeteer, why not instead use the platform accessibility apis?

**Comment by FremyCompany** [comment]
@JoelEinbinder Long story, but having access to the full platform enables quickly experimenting and testing new things which would require extra-long coordination efforts across organizations otherwise, and either prove their value or iterate on them until a more concrete proposal can be made and analyzed regarding its cost/benefit; plus it is cross-platform which is great for sharing experimental results to people regardless of their platform. 

Currently I have put that specific project on hold, because I am currently focusing on more fundamental research topics but the proof of concept I have built on top of puppeteer is very solid and I mid-term I intend to work on this again. I will say I was super impressed with the puppeteer project and API, and I say that having worked on browsers in the past, and having used all kinds of Selenium and Web Driver things including some internal-only ones at Microsoft :)

**Comment by AutoSponge** [comment]
@JoelEinbinder Maybe I misspoke. An object like `MenuListPopup` is not targetable by a developer using devtools. Therefore, it's not an object I'd be interested in for **my** purposes. As @FremyCompany said, it may have importance for other applications. Maybe I can suggest a `{ pierce: true }` option for this future method?

**Comment by JoelEinbinder** [comment]
@FremyCompany It sounds like you are doing some experimental/prototype stuff. I'd recommend copying `Accessibility.js` into your project and modifying it yourself.

@AutoSponge can you give a bit more context about your use case? Would having `elementHandle.accessibilitySnapshot` be sufficient?

**Comment by FremyCompany** [comment]
@JoelEinbinder I would consider doing that if that was required. Fwiw, your proposed `elementHandle.accessibilitySnapshot` would work really well for me (I would then navigate the dom as primary, then get the accessibility info on end nodes like input/video where the dom tree isn't providing the relevant info). If you do that,  `elementHandle.accessiblitySnapshot` should maybe have an optional parameter specifying not to return children nodes as an optimization, just the AXNode associated to the element.

**Comment by AutoSponge** [comment]
@JoelEinbinder My case: I want to find (usually interactive) DOM nodes by accessible name for testing purposes. This simultaneously tests functionality and accessibility in a black box style. Short of a new method, `queryByAXName`, I need a way to correlate the AXTree and the DOM. `this.nodeId = this._payload.nodeId` is how I'm doing it using the CDP (in the AXNode constructor). But if a better, documented approach were available, I'd scrap my hack.

**Comment by rooby** [comment]
elementHandle.accessibilitySnapshot would be great for me too as I need an easy way to test that a certain element has focus.

**Comment by AutoSponge** [comment]
@JoelEinbinder After seeing the new switch for "user agent shadow" in chrome devtools, it made me wonder if we can just use that as the flag for whether a node reference will be returned or not. Personally, I wouldn't need the user agent shadow nodes. Does that help work around some of the problems you mentioned before?

**Comment by dtinth** [comment]
I need to interact (click, type, etc.) via the accessibility tree, so I need some kind of workaround… Sharing it here:

## Workaround as of Puppeteer v5.0.0

### Getting a __raw__ accessibility tree from a Puppeteer page

```js
const client = getClient(page)
const tree = await getAccessibilityTree(client)

/**
 * @param {import("puppeteer").Page} page
 */
function getClient(page) {
  return /** @type {import('puppeteer').CDPSession} */ (
    /** @type {any} */ (page)._client
  )
}

/**
 * @param {import("puppeteer").CDPSession} client
 */
async function getAccessibilityTree(client) {
  return /** @type {import("puppeteer/lib/esm/protocol").default.Accessibility.getFullAXTreeReturnValue} */ (await client.send(
    'Accessibility.getFullAXTree',
  ))
}
```

The result will contain `nodes[*].backendDOMNodeId` which allows resolving back to a DOM node.

### Getting a DOM node given a `backendDOMNodeId`

```js
const node = await resolveNodeFromBackendNodeId(page.mainFrame(), backendDOMNodeId)

/**
 * @param {import("puppeteer").Frame} frame
 * @param {number} backendNodeId
 */
async function resolveNodeFromBackendNodeId(frame, backendNodeId) {
  const ctx = await Promise.resolve(frame.executionContext())
  return /** @type {import('puppeteer').ElementHandle} */ (
    /** @type {any} */ (ctx)._adoptBackendNodeId(backendNodeId)
  )
}
```

The result will be an element handle that we can interact with.

**Comment by kevin940726** [comment]
What's the state of this issue? I'm looking for the same feature and can't really think of any better alternatives other than accessing the private APIs for now.

My use case is to query elements based on their accessibility properties. While `aria` query could do some of the jobs, but not all of them (currently `aria` query only support querying by name and role).

If it's up to me then I would probably try to keep it simple and just return `null` for those that are not associated with a DOM element, and just let the developers to handle the rest.

**Comment by jribbens** [comment]
I'd like to know where we are with this too. I would very much like the suggested `elementHandle.accessibilitySnapshot`. The documentation claims that `page.accessibility.snapshot({ root: elementHandle })` should work, but as far as I can tell it simply doesn't, it just always returns `null`.
